File: human_data/MB04/MB04_parvo_fit_opt.py
"""MB04 metabolic cost per optimization condition, from the Parvo breath-by-breath CSV.

Each condition's Brockway metabolic rate is fit with a first-order response
`a + b*exp(-t/c)`, and the fit's mean over 240-360 s, minus the standing rate,
is the condition's estimated cost.
"""

from pathlib import Path
import pypolar as plr
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from scipy.integrate import quad
from scipy.optimize import curve_fit

HERE        = Path(__file__).parent
DATA_PATH   = HERE / 'MB04__0_20231223_0713.CSV'
OUTPUT_PATH = HERE / 'parvo_results_opt.csv'

MASS                = 89        # kg
HEADER_LINES        = 29
TRIAL_DURATION      = 180       # s of data each fit sees
STAND_DURATION      = 120       # s
STEADY_STATE_WINDOW = 60        # s at the end of a trial that 'avg' and 'std' use
ESTIMATE_WINDOW     = (240, 360)  # s the fitted curve is averaged over
MAX_RISE_TIME       = 52        # s, upper bound on c
PLOT_TRIAL          = 'walk_opt_8'

COLUMNS = ['time', 'vo2', 'vo2_kg', 'mets', 'vco2', 've', 'rer', 'rr', 'vt', 'feo2', 'feco2', 'ackcal']

# Condition start times as (minutes, seconds) from the start of the recording
STAND = (1, 1)
CONDITIONS = {
    'walk_opt_1':  (7, 59),   'walk_opt_2':  (11, 49),  'walk_opt_3':  (15, 7),
    'walk_opt_4':  (18, 27),  'walk_opt_5':  (49, 38),  'walk_opt_6':  (53, 2),
    'walk_opt_7':  (56, 46),  'walk_opt_8':  (59, 47),  'walk_opt_9':  (63, 5),
    'walk_opt_10': (66, 34),  'walk_opt_11': (69, 46),  'walk_opt_12': (73, 21),
    'walk_opt_13': (91, 42),  'walk_opt_14': (95, 4),   'walk_opt_15': (104, 41),
    'walk_opt_16': (108, 2),  'walk_opt_17': (117, 18), 'walk_opt_18': (120, 36),
    'walk_opt_19': (123, 53), 'walk_opt_20': (127, 23), 'walk_opt_21': (130, 42),
    'walk_opt_22': (134, 0),  'walk_opt_23': (137, 20), 'walk_opt_24': (140, 33),
}


def first_order(t, a, b, c):
    return a + b * np.exp(-t / c)


def read_parvo(path: Path, mass: float):
    """Breath table with time in s, VO2/VCO2 in mL/min, and a Brockway `met` column in W/kg.

    Rows after the data (the 'Max VO2' summary) have no numeric time and are dropped.
    """
    df = pd.read_csv(path, skiprows=HEADER_LINES, header=None, names=COLUMNS, skipinitialspace=True)
    df = df.apply(pd.to_numeric, errors='coerce').dropna(subset=['time']).reset_index(drop=True)
    df['vo2']  *= 1000
    df['vco2'] *= 1000
    df['time'] *= 60
    df['met']   = (0.278 * df['vo2'] + 0.075 * df['vco2']) / mass

    return df


def window(df: pd.DataFrame, start: float, end: float):
    """Rows from the breath nearest `start` through the breath nearest `end`, inclusive."""
    i0 = np.argmin(np.abs(df['time'].to_numpy() - start))
    i1 = np.argmin(np.abs(df['time'].to_numpy() - end))

    return df.iloc[i0:i1 + 1]


def fit_condition(df: pd.DataFrame, start: float):
    """Summary of one condition starting at `start` s.

    Returns a dict with the steady-state mean/std of `met`, the fit parameters (a, b, c),
    the fitted curve's mean over ESTIMATE_WINDOW, and the raw trace.
    """
    end   = start + TRIAL_DURATION
    trial = window(df, start, end)
    ss    = window(df, end - STEADY_STATE_WINDOW, end)

    t, met = trial['time'].to_numpy(), trial['met'].to_numpy()
    t      = t - t[0]
    p0     = [met[-1], met[0] - met[-1], min(60, MAX_RISE_TIME)]
    params, _ = curve_fit(first_order, t, met, p0=p0,
                          bounds=([-np.inf, -np.inf, 0], [np.inf, np.inf, MAX_RISE_TIME]))

    lo, hi   = ESTIMATE_WINDOW
    estimate = quad(first_order, lo, hi, args=tuple(params))[0] / (hi - lo)

    ss_dt = np.diff(ss['time'].to_numpy())
    return {
        'avg':         ss['met'].mean(),
        'std':         ss['met'].std(ddof=1),
        'estimate':    estimate,
        'params':      params,
        'avgssrer':    ss['rer'].mean(),
        'avgweighted': np.sum(ss_dt * ss['met'].to_numpy()[1:]) / np.sum(ss_dt),
        'time':        t,
        'data':        met,
    }


def main():
    df = read_parvo(DATA_PATH, MASS)

    stand_start  = 60 * STAND[0] + STAND[1]
    resting_rate = window(df, stand_start, stand_start + STAND_DURATION)['met'].mean()

    fits = {name: fit_condition(df, 60 * m + s) for name, (m, s) in CONDITIONS.items()}

    results = pd.DataFrame({
        'trial':    list(fits),
        'avg':      [f['avg'] - resting_rate for f in fits.values()],
        'estimate': [f['estimate'] - resting_rate for f in fits.values()],
        'std':      [f['std'] for f in fits.values()],
    })
    results.to_csv(OUTPUT_PATH, index=False)
    print(f'Resting rate: {resting_rate:.4f} W/kg')
    print(results.to_string(index=False))

    fig, ax = plt.subplots()
    ax.plot(df['rer'], '-o')
    ax.set_title('RER')

    fig, ax = plt.subplots()
    ax.bar(results['trial'], results['estimate'])
    ax.tick_params(axis='x', labelrotation=90)
    ax.set_title('MB04 Optimization')

    f = fits[PLOT_TRIAL]
    t = np.linspace(f['time'][0], f['time'][-1], 500)
    fig, ax = plt.subplots()
    ax.scatter(f['time'], f['data'], color='b', label='Raw Data')
    ax.plot(t, first_order(t, *f['params']), 'r', label='1st order fit')
    ax.axhline(f['estimate'], color='k', label='Estimated Steady-State Rate')
    ax.set_xlabel('Time (s)')
    ax.set_ylabel('Metabolic Rate (W/kg)')
    ax.legend()
    ax.set_ylim((3, None))

    path = 'human_data/MB04/met.svg'
    fig.set_size_inches((8, 3))
    plr.dress_axis(ax)
    fig.savefig(path, transparent=True)
    print(path)


if __name__ == '__main__':
    main()

File: human_data/MB04/MB04_parvo_fit_opt.m
% Parvo data processing
filePath = "D:\GaTech Dropbox\ME-DboxMgmt-Young-HIKETeam\MTJWO\MTBO_icra\MB04\MB04__0_20231223_0713.CSV";
% save_path = 
% filePath = "D:\GaTech Dropbox\ME-DboxMgmt-Young-AnkleExoTeam\Data\Combined State JW Opt\pilot_3_11\PILOT_ANKLE_00_0_20231212_1451.CSV";
mass = 89;
% mass = 70;
trialDuration = 180;
stand_duration = 120;
% opt_duration = 156;
% val_duration = 300;
steady_state_window = 60;

h = {'time', 'vo2', 'vo2_kg', 'mets', 'vco2', 've', 'rer', 'rr', 'vt', 'feo2', 'feco2', 'ackcal'};
opts = detectImportOptions(filePath, 'NumHeaderLines', 29);
trialData = readtable(filePath, opts);
trialData.Properties.VariableNames = h;

trialData.vo2 = trialData.vo2 * 1000; % mL/min
trialData.vco2 = trialData.vco2 * 1000; % mL/min
trialData.time = trialData.time * 60; % seconds

met = (0.278*trialData.vo2 + 0.075*trialData.vco2) / mass; % brockway eq
trialData = [trialData, array2table(met, 'VariableNames', {'met'})];


t = struct();

% Input times from experiment
t.STAND = duration(0,1,01);
t.walk_opt_1 = duration(0,7,59);
t.walk_opt_2 = duration(0,11,49);
t.walk_opt_3 = duration(0,15,07);
t.walk_opt_4 = duration(0,18,27);
t.walk_opt_5 = duration(0,49,38);
t.walk_opt_6 = duration(0,53,02);
t.walk_opt_7 = duration(0,56,46);
t.walk_opt_8 = duration(0,59,47);
t.walk_opt_9 = duration(0,63,05);
t.walk_opt_10 = duration(0,66,34);
t.walk_opt_11 = duration(0,69,46);
t.walk_opt_12 = duration(0,73,21);
t.walk_opt_13 = duration(0,91,42);
t.walk_opt_14 = duration(0,95,04);
t.walk_opt_15 = duration(0,104,41);
t.walk_opt_16 = duration(0,108,02);
t.walk_opt_17 = duration(0,117,18);
t.walk_opt_18 = duration(0,120,36);
t.walk_opt_19 = duration(0,123,53);
t.walk_opt_20 = duration(0,127,23);
t.walk_opt_21 = duration(0,130,42);
t.walk_opt_22 = duration(0,134,00);
t.walk_opt_23 = duration(0,137,20);
t.walk_opt_24 = duration(0,140,33);

conditions = fieldnames(t);

metStruct = struct();

figure
plot(trialData.rer, '-o')
title('RER')

conditions = conditions(contains(conditions,'walk'));

resting_start = seconds(t.STAND);
resting_end = resting_start + 120;
[~, endIdx] = min(abs(trialData.time - resting_end));
[~, startIdx] = min(abs(trialData.time - resting_start));
restingData = trialData(startIdx:endIdx, :);
restingRate = mean(restingData.met);

ft = fittype('a + b*exp(-x/c)','independent','x','coefficients',{'a','b','c'});
for ii=1:length(conditions)
    c = conditions{ii};
    startTime = seconds(t.(c));
    endTime = startTime + trialDuration;
    ssStartTime = endTime - steady_state_window;
    
    [~, endIdx] = min(abs(trialData.time - endTime));
    [~, startIdx] = min(abs(trialData.time - startTime));
    [~, ssStartIdx] = min(abs(trialData.time - ssStartTime));

    cData = trialData(startIdx:endIdx, :);
    cDataSS = trialData(ssStartIdx:endIdx, :);

    opts = fitoptions( ...
        Method='NonlinearLeastSquares', ...
        StartPoint=[cData.met(end), cData.met(1)-cData.met(end), 60], ...
        Lower=[-Inf, -Inf, 0], ...
        Upper=[Inf, Inf, 52]);  % rise time is bounded to something realistic )
    fitResult = fit(cData.time - cData.time(1), cData.met, ft, opts);
    avgEstMetPower = (1/(360 - 240)) * integral(@(t) fitResult(t), 240, 360, ArrayValued=true);

    metStruct.(c) = struct();
    metStruct.(c).avg = mean(cDataSS.met);
    metStruct.(c).estimate = avgEstMetPower;
    metStruct.(c).fit = fitResult;
    metStruct.(c).std = std(cDataSS.met);
    metStruct.(c).ssdata = cDataSS.met;
    metStruct.(c).sstime = cDataSS.time;
    metStruct.(c).data = cData.met;
    metStruct.(c).time = cData.time;
    metStruct.(c).ssduration = diff(metStruct.(c).sstime);
    metStruct.(c).avgssduration = mean(metStruct.(c).ssduration);
    metStruct.(c).avgssrer = mean(cDataSS.rer);
    
    metStruct.(c).avgweighted = sum(metStruct.(c).ssduration .* metStruct.(c).ssdata(2:end)) / sum(metStruct.(c).ssduration);
    
    % h = plot(metStruct.(c).time, metStruct.(c).data, '-o');
    % % h = plot(metStruct.(c).sstime, metStruct.(c).ssdata, '-o');
    % color = get(h, 'Color');
    % 
    % tsStartTime = find(~isnan(trialData.time));
    % tsEndTime = trialData.time(tsStartTime(end));
    % tsStartTime = trialData.time(tsStartTime(1));
    % plot([tsStartTime, tsEndTime], ones(1, 2) * metStruct.(c).avg, 'Color', color);
    % plot([tsStartTime, tsEndTime], ones(1, 2) * (metStruct.(c).avg - metStruct.(c).std), '--', 'Color', color);
    % plot([tsStartTime, tsEndTime], ones(1, 2) * (metStruct.(c).avg + metStruct.(c).std), '--', 'Color', color);
end

resultTable = table();
resultTable.trial = string(fieldnames(metStruct));
for i = 1:length(resultTable.trial)
resultTable.avg(i) = metStruct.(resultTable.trial{i}).avg - restingRate;
resultTable.estimate(i) = metStruct.(resultTable.trial{i}).estimate - restingRate;
resultTable.std(i) = metStruct.(resultTable.trial{i}).std;
end


save('parvo_results_opt.mat',"resultTable")

%% cost visualization
figure
hold on
bar(resultTable.trial, resultTable.estimate,'interpreter','none')
% errorbar(1:numel(resultTable.trial), resultTable.estimate, resultTable.std,'LineStyle', 'none', 'LineWidth', 1.5,'Color','k')
title('MB04 Optimization')


%% fit visuzalization
trial = 'walk_opt_8';

figure
hold on
scatter(metStruct.(trial).time - metStruct.(trial).time(1),  metStruct.(trial).data,'b')
plot(metStruct.(trial).fit)
yline(metStruct.(trial).estimate,'Color','k')
xlabel('Time (s)')
ylabel('Metabolic Rate (W/kg)')
legend('Raw Data','1st order fit','Estimated Steady-State Rate')